File: code_snippets/java/src/main/java/concepts/services/SignupWorkflow.java
package concepts.services;

import static usecases.utils.Utils.createUserEntry;
import static usecases.utils.Utils.sendEmailWithLink;

import dev.restate.sdk.JsonSerdes;
import dev.restate.sdk.SharedWorkflowContext;
import dev.restate.sdk.WorkflowContext;
import dev.restate.sdk.annotation.Shared;
import dev.restate.sdk.annotation.Workflow;
import dev.restate.sdk.common.DurablePromiseKey;
import dev.restate.sdk.http.vertx.RestateHttpEndpointBuilder;
import usecases.utils.User;

// <start_here>
@Workflow
public class SignupWorkflow {

  // <mark_3>
  private static final DurablePromiseKey<String> LINK_CLICKED =
      DurablePromiseKey.of("link_clicked", JsonSerdes.STRING);

  // </mark_3>

  // <mark_1>
  @Workflow
  public boolean run(WorkflowContext ctx, User user) {
    // workflow ID = user ID; workflow runs once per user
    String userId = ctx.key();

    // <mark_2>
    ctx.run(() -> createUserEntry(user));
    // </mark_2>

    // <mark_2>
    String secret = ctx.random().nextUUID().toString();
    ctx.run(() -> sendEmailWithLink(userId, user, secret));
    // </mark_2>

    // <mark_3>
    String clickSecret = ctx.promise(LINK_CLICKED).awaitable().await();
    // </mark_3>

    return clickSecret.equals(secret);
  }

  // </mark_1>

  // <mark_3>
  @Shared
  public void click(SharedWorkflowContext ctx, String secret) {
    ctx.promiseHandle(LINK_CLICKED).resolve(secret);
  }

  // </mark_3>

  public static void main(String[] args) {
    RestateHttpEndpointBuilder.builder().bind(new SignupWorkflow()).buildAndListen();
  }
}
// <end_here>

File: code_snippets/kotlin/src/main/kotlin/concepts/services/SignupWorkflow.kt
package concepts.services

import dev.restate.sdk.annotation.Shared
import dev.restate.sdk.annotation.Workflow
import dev.restate.sdk.http.vertx.RestateHttpServer
import dev.restate.sdk.kotlin.*
import dev.restate.sdk.kotlin.endpoint.endpoint
import kotlinx.serialization.Serializable

// <start_here>
@Workflow
class SignupWorkflow {

  companion object {
    // <mark_3>
    private val LINK_CLICKED = durablePromiseKey<String>("link_clicked")
    // </mark_3>
  }

  // <mark_1>
  @Workflow
  suspend fun run(ctx: WorkflowContext, user: User): Boolean {
    // workflow ID = user ID; workflow runs once per user
    val userId = ctx.key()

    // <mark_2>
    ctx.runBlock { createUserEntry(user) }
    // </mark_2>

    // <mark_2>
    val secret = ctx.random().nextUUID().toString()
    ctx.runBlock { sendEmailWithLink(userId, user, secret) }
    // </mark_2>

    // <mark_3>
    val clickSecret: String = ctx.promise(LINK_CLICKED).future().await()
    // </mark_3>

    return clickSecret == secret
  }
  // </mark_1>

  // <mark_3>
  @Shared
  suspend fun click(ctx: SharedWorkflowContext, secret: String) {
    ctx.promiseHandle(LINK_CLICKED).resolve(secret)
  }
  // </mark_3>
}

fun main() {
  RestateHttpServer.listen(endpoint { bind(SignupWorkflow()) })
}
// <end_here>

fun createUserEntry(user: User) {}

fun sendEmailWithLink(email: String?, user: User, secret: String?) {}

@Serializable data class User(val id: String, val email: String, val name: String)
